### ISSUE TITLE
Fallback syntax ?

### ISSUE THREAD

**Issue by mckellygit** [sharkdp/bat#1341]
Hi,
Thank you for bat, it is awesome (and so is fd!)
Is there a way to have a fallback syntax ?
I don't want to force a new syntax for known file types, I just want to set it to something reasonable for files that did not match any known type.
thx,
-m

**Comment by sharkdp** [comment]
Thank you for the feedback.

>  Is there a way to have a fallback syntax ?

There is no way to do this currently (maybe by overwriting the `plaintext` syntax..).

>  I don't want to force a new syntax for known file types, I just want to set it to something reasonable for files that did not match any known type.

What would "something reasonable" be?

**Comment by mckelly2833** [comment]
@sharkdp I agree what makes a good fallback syntax is just my opinion, but it could be set to whatever a user wants, including none.  I prefer some color to none though.  For me, I would probably try bash as a fallback.
I'll look into plaintext.
thx,
-m

**Comment by mckelly2833** [comment]
*I think* I have it working ok with a hacked sublime bash syntax file added as plaintext.
thx so much,
-m

**Comment by sharkdp** [comment]
Glad that worked :+1: 

I am not sure if this is something that should be added in a more official manner. I'm inclined to say no, though. It seems to me that most users would be irritated by the potentially wrong highlighting.

I'm therefore closing the issue for now, but let me know if someone feels otherwise.

**Comment by wommel** [comment]
I think it would be helpfull to have some basic highlighting for unsupported languages.

Nano in this case (besides some "Nano"  shenanigans) highlights everthing that starts with `#` as a commtent. This is basicaly correct for any kind of text based config file i could find which is neither json nor xml. This is a great help to at least conceive the basic structure of a foreign config file or quickly see the important parts of a known one.

**Comment by mmcnl** [comment]
@mckellygit do you have your hacked plaintext syntax available somewhere? I'd be interested in trying it out. Thanks!

**Comment by mckellygit** [comment]
Hi,
I created dir: ~/.config/bat/syntaxes/sublime-plaintext-syntax
 and in this dir, I added this file: plaintext.sublime-syntax
 and in this file I added these lines at the top, after the %YAML 1.2 line:
```
name: Plain Text
scope: source.shell.bash
version: 2

extends: Packages/ShellScript/commands-builtin-shell-bash.sublime-syntax

file_extensions:
  - txt
```
I forget where I found the original plaintext.sublime-syntax file but I will look again or I can attach mine here.
thx,
-m

**Comment by mckellygit** [comment]
I think the file was from:
```
https://github.com/sublimehq/Packages/blob/master/ShellScript/Bash.sublime-syntax
```
so I renamed it and editted the first few lines to look like above.

**Comment by mmcnl** [comment]
awesome, thanks @mckellygit!

**Comment by mmcnl** [comment]
@sharkdp  I am wondering if you would re-consider adding this feature as an option that requires the user to opt-in. This would be very helpful for me as a sysadmin because I frequently have the same need as mentioned [above](https://github.com/sharkdp/bat/issues/1341#issuecomment-761527586)
> Nano in this case ... highlights everything that starts with # as a comment. This is basically correct for any kind of text based config file i could find which is neither json nor xml. This is a great help to at least conceive the basic structure of a foreign config file or quickly see the important parts of a known one.

In terms of UX, it could be something like `--fallback-syntax=ini` which would only take effect if the current file doesn't match any other language

**Comment by Enselic** [comment]
To me this sounds like reasonable opt-in functionality, so if someone comes up with a good solution with a good set of integration tests, I think that is something we could add. Since @sharkdp was not sure if closing was the right move, and since there seems to be more than a one-off interest in something like this, I think we should re-open this one.

**Comment by mckellygit** [comment]
>> but that approach doesn't work for arbitrary/unknown file extensions.
It does for me.  Not sure if I explained my config correctly, but I do get syntax highlighting on arbitrary and unknown files/extensions.
But of course having a builtin fallback syntax would be awesome.

**Comment by mmcnl** [comment]
@mckellygit I went back and tried your approach again and was able to get it to work.  🎉  Thanks again for the inspiration and instructions!

In the process, I simplified the syntax definition file to only highlight line comments and ignore all of the other bash syntax rules which, in most cases for me, were not needed (eg in `monit` config files). Most likely someone with actual sublime syntax knowledge could simplify this further.

For anyone following along, the full sequence of steps to set this up is:

1. create a `~/.config/bat/syntaxes/sublime-plaintext-syntax/plaintext.sublime-syntax` file with the content below
2. rebuild the bat cache with `bat cache --build` (I was the missing this step the first time I tried)

```
# /root/.config/bat/syntaxes/sublime-plaintext-syntax/plaintext.sublime-syntax

name: Plain Text
scope: source.shell.bash
version: 2

file_extensions:
  - txt

first_line_match: |-
  (?x:
    \s*\*
  )

contexts:
  main:
    - match: ''
      push: [statements]

  statements:
    - include: comments

  comments:
    - match: \#+
      scope: punctuation.definition.comment.shell
      push: comments-body

  comments-body:
    - meta_scope: comment.line.number-sign.shell
    - match: \n
      pop: 1
```

**Comment by mckellygit** [comment]
ok, great.  I am so sorry I forgot to list the bat cache --build step.

**Comment by BattleCh1cken** [comment]
I'd like to work on this as my first rust contribution, is this still relevant?

**Comment by keith-hall** [comment]
Yep, still relevant :)

**Comment by Xavrir** [comment]
I’m working on this and plan to open a PR implementing an opt-in  flag with integration tests.

**Comment by Xavrir** [comment]
Correction: implementing opt-in `--fallback-syntax` (fallback-only; does not override `--language`) with integration tests.

**Comment by Xavrir** [comment]
Opened PR #3617 implementing  (alias: ) with integration tests and no behavior changes to .

**Comment by Xavrir** [comment]
Correction: opened PR #3617 implementing --fallback-syntax (alias: --fallback-language) with integration tests and no behavior changes to --language.

**Comment by jgarte** [comment]
Hi, is this issue complete?

**Comment by mckellygit** [comment]
Hi,

I have not tried configuring a fallback syntax and removing my custom plain-text syntax file.
I will do that and report back.
But before I even test - if this means the highlighting of unknown files happens now then I'd say yes this issue is complete.

-m
